File: mergify_cli/stack/list.py
#
#  Copyright © 2021-2026 Mergify SAS
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from __future__ import annotations

import asyncio
import dataclasses
import json
import sys
import typing

import click

from mergify_cli import console
from mergify_cli import console_error
from mergify_cli import utils
from mergify_cli.exit_codes import ExitCode
from mergify_cli.stack import changes
from mergify_cli.stack.push import LocalBranchInvalidError
from mergify_cli.stack.push import check_local_branch


StackEntryStatusT = typing.Literal["merged", "draft", "open", "no_pr"]
CIStatusT = typing.Literal["passing", "failing", "pending", "unknown"]
ReviewStatusT = typing.Literal["approved", "changes_requested", "pending", "unknown"]

_STATUS_DISPLAY: dict[StackEntryStatusT, tuple[str, str]] = {
    "merged": ("✓ merged", "purple"),
    "draft": ("● draft", "yellow"),
    "open": ("● open", "green"),
    "no_pr": ("○ no PR", "dim"),
}

_CI_STATUS_DISPLAY: dict[CIStatusT, tuple[str, str]] = {
    "passing": ("✓ passing", "green"),
    "failing": ("✗ failing", "red"),
    "pending": ("● pending", "yellow"),
    "unknown": ("—", "dim"),
}

_REVIEW_STATUS_DISPLAY: dict[ReviewStatusT, tuple[str, str]] = {
    "approved": ("✓ approved", "green"),
    "changes_requested": ("✗ changes requested", "red"),
    "pending": ("● pending", "yellow"),
    "unknown": ("—", "dim"),
}

if typing.TYPE_CHECKING:
    import httpx

    from mergify_cli import github_types


@dataclasses.dataclass
class CICheck:
    name: str
    status: str

    def to_dict(self) -> dict[str, str]:
        return {"name": self.name, "status": self.status}


@dataclasses.dataclass
class Review:
    user: str
    state: str

    def to_dict(self) -> dict[str, str]:
        return {"user": self.user, "state": self.state}


@dataclasses.dataclass
class StackListEntry:
    """A single entry in the stack list."""

    commit_sha: str
    title: str
    change_id: str
    status: StackEntryStatusT
    pull_number: int | None = None
    pull_url: str | None = None
    ci_status: CIStatusT = "unknown"
    ci_checks: list[CICheck] = dataclasses.field(default_factory=list)
    review_status: ReviewStatusT = "unknown"
    reviews: list[Review] = dataclasses.field(default_factory=list)
    mergeable: bool | None = None

    def to_dict(self) -> dict[str, typing.Any]:
        return {
            "commit_sha": self.commit_sha,
            "title": self.title,
            "change_id": self.change_id,
            "status": self.status,
            "pull_number": self.pull_number,
            "pull_url": self.pull_url,
            "ci_status": self.ci_status,
            "ci_checks": [c.to_dict() for c in self.ci_checks],
            "review_status": self.review_status,
            "reviews": [r.to_dict() for r in self.reviews],
            "mergeable": self.mergeable,
        }


@dataclasses.dataclass
class StackListOutput:
    """Output structure for the stack list command."""

    branch: str
    trunk: str
    entries: list[StackListEntry]

    def to_dict(self) -> dict[str, typing.Any]:
        return {
            "branch": self.branch,
            "trunk": self.trunk,
            "entries": [e.to_dict() for e in self.entries],
        }


def _format_ci_display(entry: StackListEntry, *, verbose: bool) -> str:
    if entry.ci_status == "unknown":
        return ""
    if verbose and entry.ci_checks:
        checks = []
        for check in entry.ci_checks:
            if check.status == "success":
                checks.append(f"[green]✓ {check.name}[/]")
            elif check.status == "failure":
                checks.append(f"[red]✗ {check.name}[/]")
            else:
                checks.append(f"[yellow]● {check.name}[/]")
        return f"CI: {', '.join(checks)}"
    text, color = _CI_STATUS_DISPLAY[entry.ci_status]
    return f"CI: [{color}]{text}[/]"


def _format_review_display(entry: StackListEntry, *, verbose: bool) -> str:
    if entry.review_status == "unknown":
        return ""
    if verbose and entry.reviews:
        reviewers = []
        for review in entry.reviews:
            if review.state == "APPROVED":
                reviewers.append(f"[green]✓ {review.user}[/]")
            elif review.state == "CHANGES_REQUESTED":
                reviewers.append(f"[red]✗ {review.user}[/]")
            else:
                reviewers.append(f"[dim]{review.user}[/]")
        return f"Review: {', '.join(reviewers)}"
    text, color = _REVIEW_STATUS_DISPLAY[entry.review_status]
    return f"Review: [{color}]{text}[/]"


def _get_entry_status(
    pull: github_types.PullRequest | None,
) -> StackEntryStatusT:
    """Determine the status of a stack entry based on its PR state."""
    if pull is None:
        return "no_pr"
    if pull["merged_at"]:
        return "merged"
    if pull["draft"]:
        return "draft"
    return "open"


def _get_status_display(status: StackEntryStatusT) -> tuple[str, str]:
    """Return the display text and color for a status."""
    return _STATUS_DISPLAY[status]


def _compute_ci_status(
    check_runs: list[dict[str, typing.Any]],
) -> tuple[CIStatusT, list[CICheck]]:
    """Compute CI status from GitHub check run data."""
    if not check_runs:
        return ("unknown", [])

    checks: list[CICheck] = []
    has_pending = False
    has_failure = False

    for run in check_runs:
        name = run.get("name", "")
        if run.get("status") != "completed":
            checks.append(CICheck(name=name, status="pending"))
            has_pending = True
        elif run.get("conclusion") in {"success", "skipped"}:
            checks.append(CICheck(name=name, status="success"))
        else:
            checks.append(CICheck(name=name, status="failure"))
            has_failure = True

    if has_failure:
        status: CIStatusT = "failing"
    elif has_pending:
        status = "pending"
    else:
        status = "passing"

    return (status, checks)


def _compute_review_status(
    reviews_data: list[dict[str, typing.Any]],
) -> tuple[ReviewStatusT, list[Review]]:
    """Compute review status from GitHub review data."""
    if not reviews_data:
        return ("unknown", [])

    # Keep latest review per user (APPROVED/CHANGES_REQUESTED/DISMISSED
    # take precedence over COMMENTED)
    latest_by_user: dict[str, str] = {}
    for review in reviews_data:
        user = review.get("user", {}).get("login", "")
        state = review.get("state", "")
        if not user:
            continue
        if (
            state in {"APPROVED", "CHANGES_REQUESTED", "DISMISSED"}
            or user not in latest_by_user
        ):
            latest_by_user[user] = state

    reviews = [Review(user=u, state=s) for u, s in latest_by_user.items()]

    has_changes_requested = any(r.state == "CHANGES_REQUESTED" for r in reviews)
    has_approved = any(r.state == "APPROVED" for r in reviews)

    if has_changes_requested:
        status: ReviewStatusT = "changes_requested"
    elif has_approved:
        status = "approved"
    else:
        status = "pending"

    return (status, reviews)


_MAX_CONCURRENT_API_CALLS = 5


async def _fetch_pr_details(
    client: httpx.AsyncClient,
    user: str,
    repo: str,
    entries: list[StackListEntry],
    pulls: dict[int, github_types.PullRequest],
) -> None:
    """Fetch CI checks and reviews for each PR and update entries in place."""
    sem = asyncio.Semaphore(_MAX_CONCURRENT_API_CALLS)

    async def _fetch_for_entry(entry: StackListEntry) -> None:
        if entry.pull_number is None:
            return

        pull = pulls.get(entry.pull_number)
        if pull is None:
            return

        head_sha = pull["head"]["sha"]

        async with sem:
            r_checks, r_reviews = await asyncio.gather(
                client.get(
                    f"/repos/{user}/{repo}/commits/{head_sha}/check-runs",
                ),
                client.get(
                    f"/repos/{user}/{repo}/pulls/{entry.pull_number}/reviews",
                ),
            )

        check_runs = r_checks.json().get("check_runs", [])
        entry.ci_status, entry.ci_checks = _compute_ci_status(check_runs)

        reviews_data = r_reviews.json()
        entry.review_status, entry.reviews = _compute_review_status(reviews_data)

    await asyncio.gather(*[_fetch_for_entry(e) for e in entries])


def display_stack_list(output: StackListOutput, *, verbose: bool = False) -> None:
    """Display the stack list in human-readable format using rich console."""
    console.print(
        f"\nStack on [cyan]{output.branch}[/] → [cyan]{output.trunk}[/]:\n",
    )

    if not output.entries:
        console.print("  No commits in stack", style="dim")
        return

    for entry in output.entries:
        status_text, status_color = _get_status_display(entry.status)
        short_sha = entry.commit_sha[:7]

        if entry.pull_number is not None:
            conflict = " [red]✗ conflicting[/]" if entry.mergeable is False else ""

            console.print(
                f"  [{status_color}]{status_text}[/] "
                f"[bold]#{entry.pull_number}[/] {entry.title} "
                f"[dim]({short_sha})[/]{conflict}",
            )

            # Status line (CI + review)
            ci_display = _format_ci_display(entry, verbose=verbose)
            review_display = _format_review_display(entry, verbose=verbose)
            parts = [p for p in [ci_display, review_display] if p]
            if parts:
                console.print(f"     {' | '.join(parts)}")

            console.print(f"     [dim]{entry.pull_url}[/]\n")
        else:
            console.print(
                f"  [{status_color}]{status_text}[/] {entry.title} "
                f"[dim]({short_sha})[/]\n",
            )


async def get_stack_list(
    github_server: str,
    token: str,
    *,
    trunk: tuple[str, str],
    branch_prefix: str | None = None,
    author: str | None = None,
    include_status: bool = True,
) -> StackListOutput:
    """Get the current stack's commits and their associated PRs.

    Args:
        github_server: GitHub API server URL
        token: GitHub personal access token
        trunk: Tuple of (remote, branch) for the trunk
        branch_prefix: Optional branch prefix for stack branches
        author: Optional author filter (defaults to token owner)

    Returns:
        StackListOutput containing branch info and list of entries
    """
    dest_branch = await utils.git_get_branch_name()

    if author is None:
        async with utils.get_github_http_client(github_server, token) as client:
            r_author = await client.get("/user")
            author = typing.cast("str", r_author.json()["login"])

    if branch_prefix is None:
        branch_prefix = await utils.get_default_branch_prefix(author)

    try:
        check_local_branch(branch_name=dest_branch, branch_prefix=branch_prefix)
    except LocalBranchInvalidError as e:
        console_error(e.message)
        console.print(
            "You should run `mergify stack list` on the branch you created in the first place",
        )
        sys.exit(ExitCode.INVALID_STATE)

    remote, base_branch = trunk

    user, repo = utils.get_slug(
        await utils.git("config", "--get", f"remote.{remote}.url"),
    )

    if base_branch == dest_branch:
        remote_url = await utils.git("remote", "get-url", remote)
        console_error(
            f"your local branch `{dest_branch}` targets itself: "
            f"`{remote}/{base_branch}` (at {remote_url}@{base_branch})",
        )
        console.print(
            "You should either fix the target branch or rename your local branch.\n\n"
            f"* To fix the target branch: `git branch {dest_branch} --set-upstream-to={remote}/main`\n"
            f"* To rename your local branch: `git branch -M {dest_branch} new-branch-name`",
        )
        sys.exit(ExitCode.INVALID_STATE)

    stack_prefix = f"{branch_prefix}/{dest_branch}" if branch_prefix else dest_branch

    base_commit_sha = await utils.git(
        "merge-base",
        "--fork-point",
        f"{remote}/{base_branch}",
    )
    if not base_commit_sha:
        console_error(
            f"common commit between `{remote}/{base_branch}` and `{dest_branch}` branches not found",
        )
        sys.exit(ExitCode.STACK_NOT_FOUND)

    async with utils.get_github_http_client(github_server, token) as client:
        remote_changes = await changes.get_remote_changes(
            client,
            user,
            repo,
            stack_prefix,
            author,
        )

        stack_changes = await changes.get_changes(
            base_commit_sha=base_commit_sha,
            stack_prefix=stack_prefix,
            base_branch=base_branch,
            dest_branch=dest_branch,
            remote_changes=remote_changes,
            only_update_existing_pulls=False,
            next_only=False,
        )

        # Build output structure
        entries: list[StackListEntry] = []
        pulls_by_number: dict[int, github_types.PullRequest] = {}
        for local_change in stack_changes.locals:
            status = _get_entry_status(local_change.pull)
            pull_number = (
                int(local_change.pull["number"]) if local_change.pull else None
            )
            entry = StackListEntry(
                commit_sha=local_change.commit_sha,
                title=local_change.title,
                change_id=local_change.id,
                status=status,
                pull_number=pull_number,
                pull_url=local_change.pull["html_url"] if local_change.pull else None,
                mergeable=local_change.pull.get("mergeable")
                if local_change.pull
                else None,
            )
            entries.append(entry)
            if pull_number is not None and local_change.pull is not None:
                pulls_by_number[pull_number] = local_change.pull

        if include_status:
            await _fetch_pr_details(client, user, repo, entries, pulls_by_number)

    return StackListOutput(
        branch=dest_branch,
        trunk=f"{remote}/{base_branch}",
        entries=entries,
    )


async def stack_list(
    github_server: str,
    token: str,
    *,
    trunk: tuple[str, str],
    branch_prefix: str | None = None,
    author: str | None = None,
    output_json: bool = False,
    verbose: bool = False,
) -> None:
    """List the current stack's commits and their associated PRs.

    Args:
        github_server: GitHub API server URL
        token: GitHub personal access token
        trunk: Tuple of (remote, branch) for the trunk
        branch_prefix: Optional branch prefix for stack branches
        author: Optional author filter (defaults to token owner)
        output_json: If True, output JSON instead of human-readable format
        verbose: If True, show detailed CI check names and reviewer names
    """
    output = await get_stack_list(
        github_server=github_server,
        token=token,
        trunk=trunk,
        branch_prefix=branch_prefix,
        author=author,
    )

    if output_json:
        click.echo(json.dumps(output.to_dict(), indent=2))
    else:
        display_stack_list(output, verbose=verbose)

File: crates/mergify-stack/src/commands/list.rs
//! `mergify stack list [--json] [--verbose]` — show the current
//! stack's commits and their associated PRs, optionally with CI
//! and review status.
//!
//! Port of `mergify_cli/stack/list.py::stack_list`. Uses the
//! shared [`crate::changes::classify`] for the per-commit
//! create/update/skip-merged/skip-up-to-date classification,
//! then optionally fans out per-PR `check-runs` and `reviews`
//! fetches the same way Python does.

use std::path::{Path, PathBuf};
use std::process::Command;

use mergify_core::CliError;
use mergify_core::HttpClient;
use serde::Serialize;
use serde_json::Value;

use crate::changes::{self, Action};
use crate::local_commits;
use crate::remote_changes;
use crate::stack_context;
use crate::trunk;

/// Per-commit summary used by `stack list` / `stack open`. The
/// shape mirrors Python's `StackListEntry.to_dict()` 1:1 so the
/// `--json` output is byte-compatible with the pre-port version.
#[derive(Debug, Clone, Serialize)]
pub struct StackListEntry {
    pub commit_sha: String,
    pub title: String,
    pub change_id: String,
    /// `"open" | "draft" | "merged" | "no_pr" | "skipped"`.
    pub status: String,
    pub pull_number: Option<u64>,
    pub pull_url: Option<String>,
    pub ci_status: String,
    pub ci_checks: Vec<CiCheck>,
    pub review_status: String,
    pub reviews: Vec<Review>,
    pub mergeable: Option<bool>,
}

#[derive(Debug, Clone, Serialize)]
pub struct CiCheck {
    pub name: String,
    pub status: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct Review {
    pub user: String,
    pub state: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct StackListOutput {
    pub branch: String,
    pub trunk: String,
    pub entries: Vec<StackListEntry>,
}

pub struct Options<'a> {
    pub repo_dir: Option<&'a Path>,
    pub client: &'a HttpClient,
    pub user: &'a str,
    pub repo: &'a str,
    pub author: &'a str,
    pub branch_prefix: &'a str,
    pub trunk: (&'a str, &'a str),
    /// When false, skip the `/check-runs` and `/reviews` fetches
    /// — `stack open` uses this since it only needs the URL.
    pub include_status: bool,
}

pub async fn run(opts: &Options<'_>) -> Result<StackListOutput, CliError> {
    let repo_dir = resolve_repo_toplevel(opts.repo_dir)?;
    let dest_branch = trunk::git_get_branch_name(Some(&repo_dir))?;
    stack_context::check_local_branch(&dest_branch, opts.branch_prefix)?;

    let (remote, base_branch) = opts.trunk;
    if base_branch == dest_branch {
        return Err(CliError::InvalidState(format!(
            "your local branch `{dest_branch}` targets itself: `{remote}/{base_branch}`. \
             Either fix the target branch (`git branch {dest_branch} \
             --set-upstream-to=<remote>/<target>`) or rename it."
        )));
    }

    let stack_prefix = if opts.branch_prefix.is_empty() {
        dest_branch.clone()
    } else {
        format!("{prefix}/{dest_branch}", prefix = opts.branch_prefix)
    };

    let trunk_ref = format!("{remote}/{base_branch}");
    let base_commit_sha =
        match run_git_capture(Some(&repo_dir), &["merge-base", "--fork-point", &trunk_ref]) {
            Ok(sha) if !sha.is_empty() => sha,
            _ => run_git_capture(Some(&repo_dir), &["merge-base", &trunk_ref, "HEAD"])?,
        };
    if base_commit_sha.is_empty() {
        return Err(CliError::StackNotFound(format!(
            "common commit between `{trunk_ref}` and `{dest_branch}` branches not found"
        )));
    }

    let remote_changes = remote_changes::get_remote_changes(
        opts.client,
        opts.user,
        opts.repo,
        &stack_prefix,
        opts.author,
    )
    .await?;
    let local = local_commits::read(&repo_dir, &base_commit_sha, "HEAD")?;
    let classified = changes::classify(&local, remote_changes)?;

    let mut entries = Vec::with_capacity(classified.locals.len());
    for local in &classified.locals {
        let (status, pull_number, pull_url, mergeable) = match &local.pull {
            None => ("no_pr".to_string(), None, None, None),
            Some(pull) => {
                let merged = pull.get("merged_at").is_some_and(|v| !v.is_null());
                let draft = pull.get("draft").and_then(Value::as_bool).unwrap_or(false);
                let status = if matches!(local.action, Action::SkipMerged) || merged {
                    "merged"
                } else if draft {
                    "draft"
                } else {
                    "open"
                };
                let pull_number = pull.get("number").and_then(Value::as_u64);
                let pull_url = pull
                    .get("html_url")
                    .and_then(Value::as_str)
                    .map(str::to_owned);
                let mergeable = pull.get("mergeable").and_then(Value::as_bool);
                (status.to_string(), pull_number, pull_url, mergeable)
            }
        };
        entries.push(StackListEntry {
            commit_sha: local.commit_sha.clone(),
            title: local.title.clone(),
            change_id: local.change_id.clone(),
            status,
            pull_number,
            pull_url,
            ci_status: "unknown".to_string(),
            ci_checks: Vec::new(),
            review_status: "unknown".to_string(),
            reviews: Vec::new(),
            mergeable,
        });
    }

    if opts.include_status {
        fetch_pr_details(
            opts.client,
            opts.user,
            opts.repo,
            &classified.locals,
            &mut entries,
        )
        .await?;
    }

    Ok(StackListOutput {
        branch: dest_branch,
        trunk: trunk_ref,
        entries,
    })
}

/// Per-PR fan-out: for each entry with a pull, fetch
/// `/check-runs` + `/reviews` and fold them into the entry's
/// `ci_*` / `review_*` fields. Sequential rather than concurrent
/// since GitHub's secondary rate limit kicks in around ~80
/// concurrent calls on the same endpoint pool — and a typical
/// stack has well under that.
async fn fetch_pr_details(
    client: &HttpClient,
    user: &str,
    repo: &str,
    classified: &[crate::changes::LocalChange],
    entries: &mut [StackListEntry],
) -> Result<(), CliError> {
    for (i, change) in classified.iter().enumerate() {
        let Some(pull) = &change.pull else { continue };
        let Some(head_sha) = pull.pointer("/head/sha").and_then(Value::as_str) else {
            continue;
        };
        let Some(pull_number) = pull.get("number").and_then(Value::as_u64) else {
            continue;
        };

        let checks_path = format!("/repos/{user}/{repo}/commits/{head_sha}/check-runs");
        let reviews_path = format!("/repos/{user}/{repo}/pulls/{pull_number}/reviews");

        let checks_payload: Value = client.get(&checks_path).await?;
        let reviews_payload: Value = client.get(&reviews_path).await?;

        let check_runs: Vec<Value> = checks_payload
            .get("check_runs")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default();
        let (ci_status, ci_checks) = compute_ci_status(&check_runs);
        entries[i].ci_status = ci_status.to_string();
        entries[i].ci_checks = ci_checks;

        let reviews: Vec<Value> = reviews_payload.as_array().cloned().unwrap_or_default();
        let (review_status, review_list) = compute_review_status(&reviews);
        entries[i].review_status = review_status.to_string();
        entries[i].reviews = review_list;
    }
    Ok(())
}

fn compute_ci_status(check_runs: &[Value]) -> (&'static str, Vec<CiCheck>) {
    if check_runs.is_empty() {
        return ("unknown", Vec::new());
    }
    let mut checks = Vec::with_capacity(check_runs.len());
    let mut has_pending = false;
    let mut has_failure = false;
    for run in check_runs {
        let name = run
            .get("name")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string();
        let status = run.get("status").and_then(Value::as_str);
        let conclusion = run.get("conclusion").and_then(Value::as_str);
        let kind = if status != Some("completed") {
            has_pending = true;
            "pending"
        } else if matches!(conclusion, Some("success" | "skipped")) {
            "success"
        } else {
            has_failure = true;
            "failure"
        };
        checks.push(CiCheck {
            name,
            status: kind.to_string(),
        });
    }
    let overall = if has_failure {
        "failing"
    } else if has_pending {
        "pending"
    } else {
        "passing"
    };
    (overall, checks)
}

fn compute_review_status(reviews: &[Value]) -> (&'static str, Vec<Review>) {
    if reviews.is_empty() {
        return ("unknown", Vec::new());
    }
    // Mirrors Python: keep the latest per user; APPROVED /
    // CHANGES_REQUESTED / DISMISSED override an earlier COMMENTED.
    let mut latest_by_user: std::collections::BTreeMap<String, String> =
        std::collections::BTreeMap::new();
    for review in reviews {
        let user = review
            .pointer("/user/login")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        if user.is_empty() {
            continue;
        }
        let state = review
            .get("state")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        let promoting = matches!(
            state.as_str(),
            "APPROVED" | "CHANGES_REQUESTED" | "DISMISSED"
        );
        if promoting || !latest_by_user.contains_key(&user) {
            latest_by_user.insert(user, state);
        }
    }
    let list: Vec<Review> = latest_by_user
        .into_iter()
        .map(|(user, state)| Review { user, state })
        .collect();
    let has_changes = list.iter().any(|r| r.state == "CHANGES_REQUESTED");
    let has_approved = list.iter().any(|r| r.state == "APPROVED");
    let overall = if has_changes {
        "changes_requested"
    } else if has_approved {
        "approved"
    } else {
        "pending"
    };
    (overall, list)
}

fn resolve_repo_toplevel(repo_dir: Option<&Path>) -> Result<PathBuf, CliError> {
    let raw = run_git_capture(repo_dir, &["rev-parse", "--show-toplevel"])?;
    Ok(PathBuf::from(raw))
}

fn run_git_capture(repo_dir: Option<&Path>, args: &[&str]) -> Result<String, CliError> {
    let mut cmd = Command::new("git");
    if let Some(dir) = repo_dir {
        cmd.arg("-C").arg(dir);
    }
    cmd.args(args);
    let output = cmd
        .output()
        .map_err(|e| CliError::Generic(format!("failed to spawn `git {}`: {e}", args.join(" "))))?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr).trim().to_string();
        return Err(CliError::Generic(if stderr.is_empty() {
            format!("`git {}` failed", args.join(" "))
        } else {
            stderr
        }));
    }
    let stdout = String::from_utf8(output.stdout).map_err(|e| {
        CliError::Generic(format!("`git {}` output is not UTF-8: {e}", args.join(" ")))
    })?;
    Ok(stdout.trim_end().to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn compute_ci_status_classifies_passing() {
        let runs = vec![
            json!({"name": "ci-1", "status": "completed", "conclusion": "success"}),
            json!({"name": "ci-2", "status": "completed", "conclusion": "skipped"}),
        ];
        let (status, checks) = compute_ci_status(&runs);
        assert_eq!(status, "passing");
        assert_eq!(checks.len(), 2);
        assert!(checks.iter().all(|c| c.status == "success"));
    }

    #[test]
    fn compute_ci_status_classifies_failing_over_pending() {
        let runs = vec![
            json!({"name": "ci-1", "status": "in_progress"}),
            json!({"name": "ci-2", "status": "completed", "conclusion": "failure"}),
        ];
        let (status, _) = compute_ci_status(&runs);
        assert_eq!(status, "failing");
    }

    #[test]
    fn compute_ci_status_classifies_pending() {
        let runs = vec![json!({"name": "ci-1", "status": "queued"})];
        let (status, _) = compute_ci_status(&runs);
        assert_eq!(status, "pending");
    }

    #[test]
    fn compute_review_status_latest_per_user_promoting_states() {
        let reviews = vec![
            json!({"user": {"login": "alice"}, "state": "COMMENTED"}),
            json!({"user": {"login": "alice"}, "state": "APPROVED"}),
            json!({"user": {"login": "bob"}, "state": "CHANGES_REQUESTED"}),
        ];
        let (status, list) = compute_review_status(&reviews);
        assert_eq!(status, "changes_requested");
        assert_eq!(list.len(), 2);
        let by_user: std::collections::HashMap<_, _> =
            list.into_iter().map(|r| (r.user, r.state)).collect();
        assert_eq!(by_user["alice"], "APPROVED");
        assert_eq!(by_user["bob"], "CHANGES_REQUESTED");
    }

    #[test]
    fn compute_review_status_approved_when_no_changes_requested() {
        let reviews = vec![json!({"user": {"login": "alice"}, "state": "APPROVED"})];
        let (status, _) = compute_review_status(&reviews);
        assert_eq!(status, "approved");
    }

    #[test]
    fn compute_review_status_pending_when_only_comments() {
        let reviews = vec![json!({"user": {"login": "alice"}, "state": "COMMENTED"})];
        let (status, _) = compute_review_status(&reviews);
        assert_eq!(status, "pending");
    }
}
